File: media_remote.go
package main

import (
	"fmt"
	"net"
	"net/http"
	"os"
	"os/exec"
	"strings"
)

const port = "9876"

func sendSpace() {
	exec.Command("osascript", "-e", `tell application "System Events" to keystroke space`).Run()
}

func volumeUp() {
	exec.Command("osascript", "-e", `set volume output volume ((output volume of (get volume settings)) + 10)`).Run()
}

func volumeDown() {
	exec.Command("osascript", "-e", `set volume output volume ((output volume of (get volume settings)) - 10)`).Run()
}

func getLocalHostname() string {
	out, err := exec.Command("scutil", "--get", "LocalHostName").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func getLocalIP() string {
	conn, err := net.Dial("udp", "8.8.8.8:80")
	if err != nil {
		return "localhost"
	}
	defer conn.Close()
	return conn.LocalAddr().(*net.UDPAddr).IP.String()
}

func handler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/html")
	w.Header().Set("Access-Control-Allow-Origin", "*")

	switch r.URL.Path {
	case "/setup":
		// Trigger permission dialogs - run this from the Mac itself
		fmt.Println("Running setup to trigger permissions...")
		exec.Command("osascript", "-e", `tell application "Google Chrome" to get name`).Run()
		exec.Command("osascript", "-e", `tell application "Safari" to get name`).Run()
		exec.Command("osascript", "-e", `tell application "System Events" to get name`).Run()
		w.Write([]byte(`<!DOCTYPE html><html><body style="font-family:-apple-system;padding:40px;">
			<h1>Setup complete</h1>
			<p>Approve any permission dialogs that appeared, then use the remote from your phone.</p>
			<p><a href="/">Go to remote</a></p>
		</body></html>`))
	case "/toggle":
		sendSpace()
		fmt.Println("Toggle received")
		w.Write([]byte("OK"))
	case "/vol-up":
		volumeUp()
		fmt.Println("Volume up")
		w.Write([]byte("OK"))
	case "/vol-down":
		volumeDown()
		fmt.Println("Volume down")
		w.Write([]byte("OK"))
	case "/":
		w.Write([]byte(`<!DOCTYPE html>
<html>
<head>
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <title>Remote</title>
    <style>
        * {
            box-sizing: border-box;
            touch-action: manipulation;
        }
        html, body {
            height: 100%;
            width: 100%;
            margin: 0;
            padding: 0;
            overflow: hidden;
        }
        body {
            font-family: -apple-system, BlinkMacSystemFont, sans-serif;
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: center;
            gap: 4vw;
            background: #000;
            padding: 20px;
        }
        button {
            font-size: 6vw;
            padding: 6vw 8vw;
            border-radius: 4vw;
            border: none;
            background: #333;
            color: white;
            cursor: pointer;
            -webkit-tap-highlight-color: transparent;
            user-select: none;
            touch-action: manipulation;
            transition: transform 0.1s ease, background 0.1s ease;
        }
        button:active,
        button.pressed {
            background: #444;
            transform: scale(0.97);
        }
        .play-pause {
            font-size: 8vw;
            padding: 12vw 16vw;
        }
        .volume-row {
            display: flex;
            gap: 4vw;
            width: 100%;
            justify-content: center;
        }
        .volume-row button {
            flex: 1;
            max-width: 40vw;
        }
    </style>
</head>
<body>
    <button class="play-pause" onclick="fetch('/toggle')">Play / Pause</button>
    <div class="volume-row">
        <button onclick="fetch('/vol-down')">Vol -</button>
        <button onclick="fetch('/vol-up')">Vol +</button>
    </div>
</body>
</html>`))
	default:
		http.NotFound(w, r)
	}
}

func main() {
	// Check if running as daemon
	isDaemon := len(os.Args) > 1 && os.Args[1] == "--daemon"

	if !isDaemon {
		// Fork to background and exit
		hostname := getLocalHostname()
		ip := getLocalIP()

		cmd := exec.Command(os.Args[0], "--daemon")
		cmd.Start()

		fmt.Println("Media Remote started in background")
		if hostname != "" {
			fmt.Printf("URL: http://%s.local:%s\n", hostname, port)
		}
		fmt.Printf("URL: http://%s:%s\n", ip, port)
		fmt.Println("\nStop with: pkill media-remote")
		return
	}

	// Running as daemon - start server silently
	http.HandleFunc("/", handler)
	http.ListenAndServe(":"+port, nil)
}

File: media_remote.py
#!/usr/bin/env python3
"""Simple media remote control server - control browser playback from your phone."""

import subprocess
from http.server import HTTPServer, BaseHTTPRequestHandler
import socket

def get_local_ip():
    """Get the local IP address for this machine."""
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("8.8.8.8", 80))
        return s.getsockname()[0]
    finally:
        s.close()

def send_space():
    """Send spacebar keystroke to the frontmost app using AppleScript."""
    script = 'tell application "System Events" to keystroke space'
    subprocess.run(["osascript", "-e", script])

class RemoteHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/toggle" or self.path == "/":
            send_space()
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            self.wfile.write(b"""
                <html><body style="font-family: -apple-system; display: flex;
                justify-content: center; align-items: center; height: 100vh; margin: 0;">
                <button onclick="fetch('/toggle')" style="font-size: 48px; padding: 40px 80px;
                border-radius: 20px; border: none; background: #007AFF; color: white;">
                Play/Pause</button>
                </body></html>
            """)
        else:
            self.send_response(404)
            self.end_headers()

    def log_message(self, format, *args):
        print(f"Request: {args[0]}")

if __name__ == "__main__":
    port = 8080
    ip = get_local_ip()
    print(f"\n🎬 Media Remote Server Running!")
    print(f"\nOpen this URL on your iPhone:")
    print(f"  http://{ip}:{port}")
    print(f"\nOr use this in an iOS Shortcut:")
    print(f"  http://{ip}:{port}/toggle")
    print(f"\nPress Ctrl+C to stop\n")

    server = HTTPServer(("0.0.0.0", port), RemoteHandler)
    server.serve_forever()
